apps/dsb-message-broker/src/utils/topic.schema.service.ts

```diff
@@ -1,53 +1,137 @@
+            } catch (error) {
+                throw new BadRequestException(['Payload is not in JSON format', error.message]);
+            }
+
+            isValid = validator.validate(payload);
+            error = JSON.stringify(validator.validate.errors);
+        }
+
+        return { isValid, error };
+    }
+
+    private async setValidator(
+        fqcn: string,
+        topic: string,
+        schemaType: string,
+        schema: JSONSchemaType<any> | any
+    ): Promise<{ validator: any; error: any }> {
         let validators = this._validators.get(fqcn);
         if (!validators) validators = {};
+        let validator, error;
 
-        if (!validators[topic]) {
-            const channel = this.addressbook.getChannel(fqcn);
+        if (schemaType === 'XSD') {
+            const isWellformed = libxml.loadXmlFromString(schema);
 
-            const schema = channel?.topics?.find(
-                (_topic: any) => _topic.namespace === topic
-            )?.schema;
+            const pathToFile = await this.saveSchemaToFile(`${topic}.${fqcn}.xsd`, schema);
 
-            if (schema) {
-                const JSONSchema: JSONSchemaType<any> = JSON.parse(schema);
-                validators[topic] = this.ajv.compile(JSONSchema);
-                this._validators.set(fqcn, validators);
+            if (isWellformed) validator = new XMLValidator(schema, pathToFile);
+            else error = JSON.stringify(libxml.wellformedErrors);
+        } else {
+            try {
+                const validate = this.ajv.compile(schema);
+                validator = new JSONValidator(validate);
+            } catch (err) {
+                error = JSON.stringify([err.message]);
             }
         }
 
-        if (validators[topic]) {
-            const _validate = validators[topic];
-            const isValid = _validate(JSON.parse(payload));
-            return { isValid, error: JSON.stringify(_validate.errors) };
-        }
+        validators[topic] = validator;
+        this._validators.set(fqcn, validators);
 
-        return { isValid: true, error: null };
+        return { validator, error };
+    }
+    private async getValidator(fqcn: string, topic: string) {
+        const validators = this._validators.get(fqcn);
+        if (validators && validators[topic]) return validators[topic];
+
+        const channel = this.addressbook.getChannel(fqcn);
+        const _topic = channel?.topics?.find((_topic: any) => _topic.namespace === topic);
+        if (!_topic || !_topic.schema) return null;
+
+        const { validator } = await this.setValidator(
+            fqcn,
+            topic,
+            _topic.schemaType,
+            _topic.schema
+        );
+        return validator;
+    }
+
+    private saveSchemaToFile(fileName: string, schema: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            const pathStr = path.resolve(__dirname, `../../../xsd_files/${fileName}`);
```

Comment:
line 29 already have the path, please use getter to or suggested above `os.tmpdir()`

Reply:
fixed as well